pkg/scheduler/plugins/minruntime/minruntime.go

```diff
@@ -0,0 +1,239 @@
+	}
+	if plugin.reclaimResolveMethod == "" {
+		plugin.reclaimResolveMethod = resolveMethodLCA
+	}
+	// setup caches on plugin init, but they will be reset on session open anyway
+	plugin.preemptProtectionCache = make(map[common_info.PodGroupID]map[common_info.PodGroupID]bool)
+	plugin.reclaimProtectionCache = make(map[common_info.PodGroupID]map[common_info.PodGroupID]bool)
+	plugin.preemptMinRuntimeCache = make(map[common_info.QueueID]metav1.Duration)
+	plugin.reclaimMinRuntimeCache = make(map[common_info.QueueID]map[common_info.QueueID]metav1.Duration)
+	return plugin
+}
+
+func (mr *minruntimePlugin) Name() string {
+	return "minruntime"
+}
+
+func (mr *minruntimePlugin) OnSessionOpen(ssn *framework.Session) {
+	ssn.AddReclaimVictimFilterFn(mr.reclaimFilterFn)
+	ssn.AddPreemptVictimFilterFn(mr.preemptFilterFn)
+	ssn.AddReclaimScenarioValidatorFn(mr.reclaimScenarioValidatorFn)
+	ssn.AddPreemptScenarioValidatorFn(mr.preemptScenarioValidatorFn)
+	mr.queues = ssn.Queues
+	// setup caches on session open
+	mr.preemptProtectionCache = make(map[common_info.PodGroupID]map[common_info.PodGroupID]bool)
+	mr.reclaimProtectionCache = make(map[common_info.PodGroupID]map[common_info.PodGroupID]bool)
+	mr.preemptMinRuntimeCache = make(map[common_info.QueueID]metav1.Duration)
+	mr.reclaimMinRuntimeCache = make(map[common_info.QueueID]map[common_info.QueueID]metav1.Duration)
+}
+
+func (mr *minruntimePlugin) OnSessionClose(ssn *framework.Session) {
+	mr.queues = nil
+	mr.preemptProtectionCache = nil
+	mr.reclaimProtectionCache = nil
+	mr.preemptMinRuntimeCache = nil
+	mr.reclaimMinRuntimeCache = nil
+}
+
+func (mr *minruntimePlugin) reclaimFilterFn(pendingJob *podgroup_info.PodGroupInfo, victim *podgroup_info.PodGroupInfo) bool {
+	protected := mr.isReclaimMinRuntimeProtected(pendingJob, victim)
+	// always return true for elastic jobs, but cache the result for scenario validator
+	log.InfraLogger.V(3).Infof("preemptFilterFn: pendingJob: %s, victim: %s, protected: %t", pendingJob.Name, victim.Name, protected)
+	if victim.MinAvailable < int32(len(victim.PodInfos)) {
+		return true
+	}
+
+	return !protected
+}
+
+func (mr *minruntimePlugin) preemptFilterFn(pendingJob *podgroup_info.PodGroupInfo, victim *podgroup_info.PodGroupInfo) bool {
+	protected := mr.isPreemptMinRuntimeProtected(pendingJob, victim)
+	// always return true for elastic jobs, but cache the result for scenario validator
+	log.InfraLogger.V(3).Infof("preemptFilterFn: pendingJob: %s, victim: %s, protected: %t", pendingJob.Name, victim.Name, protected)
+	if victim.MinAvailable < int32(len(victim.PodInfos)) {
+		return true
+	}
+	return !protected
+}
+
+func (mr *minruntimePlugin) reclaimScenarioValidatorFn(reclaimer *podgroup_info.PodGroupInfo, victims []*podgroup_info.PodGroupInfo, tasks []*pod_info.PodInfo) bool {
+	for _, victim := range victims {
+		protected := mr.isReclaimMinRuntimeProtected(reclaimer, victim)
+		if protected {
```

Comment:
Try to avoid nesting logic as much as possible:
```suggestion
		if !protected {
			continue
		}
```